--- a/src/x86/ffi64.c
+++ b/src/x86/ffi64.c
@@ -458,7 +458,7 @@
   size_t bytes, n, rtype_size;
   ffi_type *rtype;
 
-#ifndef __ILP32__
+#if !defined(__ILP32__) && !defined(__PROSPERO__)
   if (cif->abi == FFI_EFI64 || cif->abi == FFI_GNUW64)
     return ffi_prep_cif_machdep_efi64(cif);
 #endif
@@ -1155,7 +1155,7 @@
         }
     }
 
-#ifndef __ILP32__
+#if !defined(__ILP32__) && !defined(__PROSPERO__)
   if (cif->abi == FFI_EFI64 || cif->abi == FFI_GNUW64)
     {
       ffi_call_efi64(cif, fn, rvalue, avalue);
@@ -1177,7 +1177,7 @@
 ffi_call_go (ffi_cif *cif, void (*fn)(void), void *rvalue,
 	     void **avalue, void *closure)
 {
-#ifndef __ILP32__
+#if !defined(__ILP32__) && !defined(__PROSPERO__)
   if (cif->abi == FFI_EFI64 || cif->abi == FFI_GNUW64)
     {
       ffi_call_go_efi64(cif, fn, rvalue, avalue, closure);
@@ -1225,7 +1225,7 @@
   void (*dest)(void);
   char *tramp = closure->tramp;
 
-#ifndef __ILP32__
+#if !defined(__ILP32__) && !defined(__PROSPERO__)
   if (cif->abi == FFI_EFI64 || cif->abi == FFI_GNUW64)
     return ffi_prep_closure_loc_efi64(closure, cif, fun, user_data, codeloc);
 #endif
@@ -1378,7 +1378,7 @@
 ffi_prep_go_closure (ffi_go_closure* closure, ffi_cif* cif,
 		     void (*fun)(ffi_cif*, void*, void**, void*))
 {
-#ifndef __ILP32__
+#if !defined(__ILP32__) && !defined(__PROSPERO__)
   if (cif->abi == FFI_EFI64 || cif->abi == FFI_GNUW64)
     return ffi_prep_go_closure_efi64(closure, cif, fun);
 #endif
--- a/configure.host
+++ b/configure.host
@@ -321,7 +321,7 @@
 	fi
 	;;
   X86_64)
-	if test x"$TARGET_X32" = xyes; then
+	if test x"$TARGET_X32" = xyes || test "$host" = x86_64-sie-freebsd; then
 		SOURCES="ffi64.c unix64.S"
 	else
 		SOURCES="ffi64.c unix64.S ffiw64.c win64.S"
